pkg/store/proxy.go

```diff
@@ -344,11 +398,55 @@ func startStreamSeriesSet(
 
 	wg.Add(1)
 	go func() {
+		isFirstRecv := true
 		defer wg.Done()
 		defer close(s.recvCh)
-
 		for {
-			r, err := s.stream.Recv()
+			var cancel context.CancelFunc
+			if s.responseTimeout != 0 {
+				ctx, cancel = context.WithTimeout(ctx, s.responseTimeout)
+				defer cancel()
+			}
+			rCh := make(chan *recvResponse, 1)
+			var rr *recvResponse
+			go func() {
+				t0 := time.Now()
+				r, err := s.stream.Recv()
+				if isFirstRecv {
+					if r.Size() > 0 {
+						metrics.withPayload.Observe(time.Since(t0).Seconds())
+					} else {
+						metrics.withoutPayload.Observe(time.Since(t0).Seconds())
```

Comment:
Can we just skip the recv without payload? We care about time to first byte IMO

Reply:
We observe strange data from this metrics (we set 10s timeout):
![Снимок экрана 2019-11-27 в 17 44 21](https://user-images.githubusercontent.com/2960110/69733241-18cdb380-113e-11ea-9054-2244e1c13124.png)
Therefore, we want to save without_payload metrics, for investigate problems in future.

Reply:
Wait, this is because it's EOF which we should filter out. This the normal response when the server closes the stream, it should be filtered out. This is also received when context is canceled or timeout was triggered that's why you see 10s I think.

Reply:
Sorry, I don't understand... I think that we must to see minimal RT for first byte without payload...

Reply:
No, I EOF is when server closes a stream, but we account that as `first byte without payload` - I believe that's wrong

Reply:
I think that `without_payload` == `EOF on first byte` and it's represetnatine metric...

Reply:
Could you add a check for `err` here before sending it back? It probably is equal to `io.EOF` in such case and it should be filtered out.

Reply:
I don't see a reasons to check on some errors here. All errors processed out of this goroutine. This goroutine needed only for get data async.

Reply:
I also think we don't need `without_payload` metrics, let's just have time_to_first_byte metrics. What are you going to differently when you see `without_payload`  vs `with_payload` metrics increasing?

Reply:
I agree here. This metric roughly says "time until a node tells us that it doesn't have any metrics". In what cases do you think it could be useful? `payload` is a bit misleading as well, IMHO. Payload is "the actual information or message in transmitted data" and even if we get EOF it still is some kind of data. Even if we will keep this around I'd suggest renaming this to perhaps `response_kind` that could be `no_metrics` or `metrics`.